lms/djangoapps/courseware/views/index.py

```diff
@@ -152,34 +148,25 @@ def get(self, request, course_id, chapter=None, section=None, position=None):
 
                 self.view = STUDENT_VIEW
 
-                # Do the enrollment check if enable_unenrolled_access is not enabled.
                 self.course = get_course_with_access(
                     request.user, 'load', self.course_key,
                     depth=CONTENT_DEPTH,
-                    check_if_enrolled=not self.enable_unenrolled_access,
+                    check_if_enrolled=True,
+                    check_if_authenticated=True
                 )
                 self.course_overview = CourseOverview.get_from_id(self.course.id)
+                self.is_staff = has_access(request.user, 'staff', self.course)
 
-                if self.enable_unenrolled_access:
-                    # Check if the user is considered enrolled (i.e. is an enrolled learner or staff).
-                    try:
-                        check_course_access(
```

Comment:
So this is funny.  In the old code, if the flag IS on (see 163 above), we call check_course_access _again_  (very inefficient), except this time we flip the `check_if_enrolled` flag on.  We go through this whole process to know whether or not we should show the PUBLIC_VIEW or the STUDENT_VIEW.  

In the old code, this would suddenly cause `check_course_access` to throw a redirect because it didn't know about the unenrolled access flag.  Knowing that this is the only thing we changed between the first call to `check_course_access` and this one, if we suddenly throw an exception, it must be because of what we changed, so we know that the CourseAccessRedirect we just caught wasn't for some other reason.

If the course_visibility is public, then we show the public view.  Otherwise we just re-raise that exception, which we know to be enrollment-related.